src/main/java/com/google/collegeplanner/servlets/PlannerServlet.java

```diff
@@ -29,35 +35,199 @@
-    String[][] courses = {strClasses.split(",")};
+    JSONArray selectedClasses = (JSONArray) body.get("selectedClasses");
+
+    constructGraphs(selectedClasses);
+
+    ArrayList<ArrayList<String>> semesters = new ArrayList<>();
+    for (int i = numSemesters; i > 0; i--) {
+      int avgCredits = totalCredits / i;
+      ArrayList<String> semester = new ArrayList<>();
+      for (int j = 0; j < courseList.size(); j++) {
+        String course = courseList.get(j);
+        if (indegree.get(course) == 0) {
+          // Check if all corequisites also have an indegree of 0
+          if (isValidToAdd(course)) {
+            // If valid, add all to semester
+            courseList.remove(j);
+            j--;
+            avgCredits -= credits.get(course);
+            totalCredits -= credits.get(course);
+            semester.add(course);
+            for (String coreq : corequisites.get(course)) {
+              courseList.remove(coreq);
+              avgCredits -= credits.get(coreq);
+              totalCredits -= credits.get(coreq);
+              semester.add(coreq);
+            }
+          }
+        }
+        // Breaks out if there are no more required classes and we exceeded our credit limit
+        if (depths.get(course) != i && avgCredits <= 0) {
+          break;
+        }
+      }
+      // Update indegrees for all next courses
+      for (String course : semester) {
+        for (String nextCourse : nextCourses.get(course)) {
+          indegree.replace(nextCourse, indegree.get(nextCourse) - 1);
+        }
+      }
+      semesters.add(semester);
+    }
+    // If courseList is not empty, scheduling was not possible
+    if (!courseList.isEmpty()) {
+      respondWithError(
+          "Not enough semesters for schedule.", HttpServletResponse.SC_BAD_REQUEST, response);
+    }
     response.setContentType("applications/json;");
-    response.getWriter().println(new Gson().toJson(courses));
+    response.getWriter().println(new Gson().toJson(semesters));
+  }
+
+  /**
+   * Given a course, returns whether all of its corequisites have a indegree of 0
+   */
+  private boolean isValidToAdd(String course) {
+    for (String coreq : corequisites.get(course)) {
+      if (indegree.get(coreq) != 0) {
+        return false;
+      }
+    }
+    return true;
   }
 
-  private static JSONObject getBody(HttpServletRequest request)
+  /**
+   * Gets the JSON Representation of the body of the POST request
+   */
+  private JSONObject getBody(HttpServletRequest request)
       throws IOException, ParseException, NullPointerException {
     String strBody =
         request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
     JSONParser parser = new JSONParser();
     return (JSONObject) parser.parse(strBody);
   }
 
+  /**
+   * Creates the representation for the graph and reads/stores necessary information from
+   * selectedClasses
+   */
+  private void constructGraphs(JSONArray selectedClasses)
```

Comment:
All the java functions you're adding should have a brief description of each parameter